File: Java/src/com/dbc/algorithm/DynamicProgramming/PalindromePartitioningII.java
package com.dbc.algorithm.DynamicProgramming;

import java.util.Arrays;

public class PalindromePartitioningII {
    public int minCut(String s) {
        int n = s.length();
        boolean[][] remain = new boolean[n][n];
        for (int i = 0; i < n; i++) {
            Arrays.fill(remain[i], true);
        }
        for (int i = n - 1; i > -1; i--) {
            for (int j = i + 1; j < n; j++) {
                remain[i][j] = (s.charAt(i) == s.charAt(j)) && remain[i + 1][j - 1];
            }
        }

        int[] result = new int[n];
        Arrays.fill(result, Integer.MAX_VALUE);
        for (int i = 0; i < n; i++) {
            if (remain[0][i]) {
                result[i] = 0;
            } else {
                for (int j = 0; j < i; j++) {
                    if (remain[j + 1][i]) {
                        result[i] = Math.min(result[i], result[j] + 1);
                    }
                }
            }
        }
        return result[n - 1];
    }
}

File: Python/dynamic-programming/palindrome_partitioning_ii.py
def minCut(s: str) -> int:
    n = len(s)
    g = [[True] * n for _ in range(n)]
    for i in range(n - 1, -1, -1):
        for j in range(i + 1, n):
            g[i][j] = (s[i] == s[j]) and g[i + 1][j - 1]
    
    f = [float("inf")] * n
    for i in range(n):
        if g[0][i]:
            f[i] = 0
        else:
            for j in range(i):
                if g[j + 1][i]:
                    f[i] = min(f[i], f[j] + 1)
    return f[n - 1]